Identify Pod's OS during API Server admission (#2803)
kubernetes/enhancements: keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The main goal of this enhancement is to authoritatively identify Windows
pods during the API Server admission time. 

## Motivation

> Identifying Windows pods during the API Server admission time is crucial to apply

Reviewer: There are other non security constraint/context related scenarios that we could provide a better user experience around if we were able to validate per-OS specific settings during admission time. 
Should we call some of those out here too?
 
https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#storage lists some

appropriate security constraints to the pod. Without proper identification, some admission
plugins may apply unnecessary security constraints to the Windows pods or in the
worst case, don't apply security constraints at all.

### Goals

> Admission plugins and/or Admission Webhooks identifying if the incoming pod

Reviewer: We should clarify which admission plugins/webhooks will be able to act on the target OS for the incoming pod.

Author: At the time of writing this KEP, I was thinking either [runtimeclass plugin](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/runtimeclass/admission.go) or [pod security](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/security/podsecurity/admission.go) will use this identification mechanism to detect Windows pods and act up on them. The action could be rejecting those pods or an event which I elaborated  [here](https://github.com/kubernetes/enhancements/pull/2803/files#diff-0bd86b6b4704c4e10a453c7f15e74e31aacb53891087cf069ae11b8bf5293a30R139)

is Windows or not.

### Non-Goals

- Validating if the Windows pods have appropriate security constraints.
- Interaction with container runtimes.

## Proposal

As mentioned [earlier](#Motivation) identifying the Windows pods during the scheduling phase can be done
in the following ways:
- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
  Windows node specific labels.
- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec

> The problem with using nodeSelector and tolerations is that any unprivileged user/entity can apply the 

Reviewer: Another advantage of using runtime classes was that the container runtime performs validation that the correct runtime classes were used for the container on the node. This means that even if users tried to fool/bypass policy enforcement through clever usage of labels/taints/ect the containers wouldn't start.

Author: Agreed but that is after the pod lands on the node and the scope of this KEP is to identify pods at the API server admission time.

nodeSelector and/or tolerations in the pod spec making it unsecure whereas RuntimeClasses are [recommended](https://kubernetes.io/docs/concepts/containers/runtime-class/#setup) to be
created by cluster administrator making them authoritative enough to be used.

### User Stories (Optional)

#### Story 1

As a Kubernetes cluster administrator, I want appropriate security contexts to be 
applied to my Windows Pods along with Linux pods

#### Story 2

As a Kubernetes cluster administrator, I want to use my own admission webhook for
Windows pods.

> ### Notes/Constraints/Caveats (Optional)

Reviewer: We should note that operators/users would be forced to use runtime classes in all of their podspecs (or at least for all the ones targeting Windows nodes/containers) we should call that out here.

Reviewer: And also define what would happen if someone tries to schedule a pod with just a`kubernetes.io\os=windows` nodeSelector (with no runtime class specified)

Author: > We should note that operators/users would be forced to use runtime classes in all of their podspecs (or at least for all the ones targeting Windows nodes/containers) we should call that out here.

Not in the initial release but in the subsequent releases.

> And also define what would happen if someone tries to schedule a pod with just akubernetes.io\os=windows nodeSelector (with no runtime class specified)

We'd emit a warning just like how deprecated api usage would cause warning in first release. In the subsequent releases, we'd deny pod admission if it just has `kubernetes.io/os=Windows` nodeSelector as mentioned [here](https://github.com/kubernetes/enhancements/pull/2803/files#diff-0bd86b6b4704c4e10a453c7f15e74e31aacb53891087cf069ae11b8bf5293a30R172)

Reviewer 2: typo: `kubernetes.io/os=windows`

> In the subsequent releases, we'd deny pod admission if it just has `kubernetes.io/os=Windows` nodeSelector as mentioned [here](https://github.com/kubernetes/enhancements/pull/2803/files#diff-0bd86b6b4704c4e10a453c7f15e74e31aacb53891087cf069ae11b8bf5293a30R172)

that seems like a significant compatibility break... I don't anticipate pods like that getting rejected in the future

While the `scheduling` field in the RuntimeClass is created to handle scheduling constraints, we have a reserved label

> [`kubernetes.io/os`](https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetes-io-os) to identify

Reviewer: I thought it was possible to overwrite the 'kubernetes.io/os` label of nodes.

Windows nodes from their Linux counterparts. If the RuntimClass's scheduling field has a nodeSelector  `kubernetes.io/os: 'windows'`,

> we can authoritatively say that the pod is indeed a Windows one during the api-server admission time. Querying the RuntimeClass

Reviewer 3: This proposal came up in the context of Pod Security admission. The concern with using scheduling constraints to enforce admission policy is that the scheduler can be bypassed, for example by creating a pod with `NodeName` already set, or issuing a `bind` request manually (with the correct permissions).

To be secure, I'd like the Kubelet to recognize the same definitive source-of-truth that the admission controllers are using. In other words, if the admission decision is based on the nodeSelector, then the Kubelet should also verify the node selector OS, and reject the pod if there's a mismatch.

Author: > This proposal came up in the context of Pod Security admission. The concern with using scheduling constraints to enforce admission policy is that the scheduler can be bypassed, for example by creating a pod with NodeName already set, or issuing a bind request manually (with the correct permissions).

That's a good point. We can make kubelet follow the same logic during admission time. As of now, we have logic in kubelet which strips certain security constraints based on OS. I'll add some text which includes changes to kubelet soon.

Reviewer: > To be secure, I'd like the Kubelet to recognize the same definitive source-of-truth that the admission controllers are using. In other words, if the admission decision is based on the nodeSelector, then the Kubelet should also verify the node selector OS, and reject the pod if there's a mismatch.

Great idea

Reviewer: @derekwaynecarr / @dchen1107 here is another proposed kubelet behavior.
I like the idea of adding in a kubelet check to make sure the current GOOS matches `kubernetes.io/os` nodeSelector (if set for the pod) and fail if they don't match.
Any objections to adding this check?

Reviewer 4: i think that makes sense to add, but i also am fine with not needing the selector at all in favor of a desiredOS attribute.

> during admission time may be expensive operation as the admission plugin is not stateless anymore. 

Reviewer 5: How are you going to mitigate or quantify this expense?

Author: It will be amortized if we go with runtimeclass admission plugin as it is making an apicall to get , if not it'll be O(1) per pod(the admission works per pod) as pod spec can have maximum of 1 runtimeclass * No of pod creations in the cluster.(Need to check if we allow updating runtimeclasses which I think we don't do because pod spec is immutable).

Reviewer 5: I suggest calling it out here

Reviewer 3: Scheduling constraints on the RuntimeClass are merged into the PodSpec during admission (mutating), so I don't think an admission controller would actually need to query the runtimeclass.

Author: That's why I said, if we go with runtimeclass admission plugin we won't need querying as it is already making an api call. However, we'd also want the same mechanism to be used in pod security admission plugin to identify Windows pods and for that we will need to query runtimeclasses unless you're suggesting something different.

Reviewer 3: What I mean is, validating admission happens after mutating admission, so as long as you're using the runtimeclass admission plugin (which is expected, if using RuntimeClasses) then the information will already be merged into the pod by the time the PodSecurity admission plugin runs.

Author: To be clear, if we go with this approach, the runtimeclasses admission plugin is not only going to mutate, it is also going to do a validation of the pod.

Sample [PR](https://github.com/kubernetes/kubernetes/pull/103784/files) which does this.

Reviewer 2: The responsibilities and actors in this doc are hard to follow... I *think* this is proposing the following:

Determining windows pods in validating admission plugins:

1. make pod `spec.nodeSelector` `kubernetes.io/os: windows` definitive for determining a pod is a Windows pod
2. make kubelet refuse to run pods that have a `kubernetes.io/os` spec.nodeSelector that does not match the kubelet's runtime.GOOS

Mechanisms for setting spec.nodeSelector:

1. user can set spec.nodeSelector in workload template
2. user can set `spec.runtimeClassName` in workload template, admin can set `scheduling:{nodeSelector:{"kubernetes.io/os":"windows"}}` in runtimeclass, runtimeclass mutating admission copies into pod `spec.nodeSelector` during mutating admission, ensuring validating admission plugins have the pod spec.nodeSelector data available to inspect

Distinguishing the "what identifies a windows pod?" details from the "how does that data get set in the pod API?" details will make this proposal easier to reason about.

Reviewer 2: One implication of having the runtimeclass admission plugin apply the `kubernetes.io/os: windows` nodeSelector at pod creation time is that it leaves the workload resource pod templates without a `kubernetes.io/os: windows` nodeSelector. That means that admission plugins that inspect the pod templates in those workload resources will not know what OS the resulting pods will target.

As a concrete example, if you set up PodSecurityAdmission to warn users if their deployments don't fit inside the `restricted` Pod Security Standards level, and their deployment targeted a windows runtimeClassName but didn't set a `kubernetes.io/os: windows` nodeSelector and didn't set `allowPrivilegeEscalation: false`, the user would get warned that the pod template in the deployment didn't fit within the `restricted` policy level.

Reviewer 2: > > One implication of having the runtimeclass admission plugin apply the kubernetes.io/os: windows nodeSelector at pod creation time is that it leaves the workload resource pod templates without a kubernetes.io/os: windows nodeSelector.
> 
> That is good point. Isn't that an existing problem with runtimeclass admission plugin

workload objects targeting a windows runtimeClassName which lack a windows node selector aren't an issue because nothing in in-tree admission uses node selectors as an input for warning/validating the pod template.

> * Have runtimeclass admission plugin also update the workload template's nodeSelector and tolerations instead of just updating the pod object so that the validation can be smooth in PodSecurityAdmission plugin

I don't think this is a good idea. Pod object selectors are immutable once set... it is significantly harder to correctly merge user- and admission-specified selectors on objects which can be updated.

> * Have the same logic of querying runtimeclasses for the workload templates/pods in PodSecurityAdmission plugin and instead warn for both workload templates/pods.

I also don't think PodSecurity admission depending on current state of RuntimeClass objects is a good idea... that admission plugin works hard to keep the evaluation of a given policy level against a given pod not depend on the state of anything else in the cluster.

Author: > workload objects targeting a windows runtimeClassName which lack a windows node selector aren't an issue because nothing in in-tree admission uses node selectors as an input for warning/validating the pod template.

Correct. That's why I said compounded by validating presence `kubernetes.io/os` label. 

> I also don't think PodSecurity admission depending on current state of RuntimeClass objects is a good idea... that admission plugin works hard to keep the evaluation of a given policy level against a given pod not depend on the state of anything else in the cluster.

I agree with the first part that it is significantly hard to update Pod object selectors in a mutating admission plugin however if PodSecurity admission plugin is anyways allowing certain runtimeclasses to excluded statically via config, how difficult would it to be to do the same dynamically. I know that we prefer admission plugins to be stateless however since PodSecurity admission plugin is not just validating pods and it is validating workload templates and we cannot update the Pod templates in workload spec, I believe there are no alternatives to having the logic in PodSecurity admission plugin.

Author: Had an offline discussion with @liggitt. The main points are:
- Have the nodeSelector explicitly set by end-user in the workload templates so that PodSecurityAdmission plugin can ignore workloads that have windows nodeSelector set.
- Let kubelet check for the nodeSelector and see if it matches node.Labels[` kubernetes.io/os `] instead of checking `runtime.GOOS` directly.

I am fine with the approach suggested. The only downside I can think of above is since nodeSelector can be added by anyone(not governed by RBAC or quota), can it be trusted during API admission time?

Reviewer 2: The selector is as trustworthy as the labels on the node are today. The two actors that pay attention to the selector are the scheduler and the kubelet. If the change to the kubelet to additionally verify the selector against labels forced to match the actual OS of the node, that closes the only gap I can see (if the Node API object label did not match the node OS)

Author: Won't we also make PodSecurityAdmission plugin trust the nodeSelector label during API admission time to differentiate Windows workloads from Linux?

Reviewer 2: once we know the node will refuse to run a pod with an OS selector that does not match the actual OS, that seems ok

Reviewer 4: there is a symmetry issue we should call out:

>make pod spec.nodeSelector kubernetes.io/os: windows definitive for determining a pod is a Windows pod

it is also authoritative to determine if a pod is on a linux host.

for a broader ecosystem of webhooks/validation/injectors, is it easier to parse pod.spec.nodeSelector or easier to just have a dedicated field that says this pod requires this OS.

### Risks and Mitigations

Existing users may be impacted as we're enforcing to use RuntimeClasses instead of plain nodeSelector and tolerations.

The downside of this approach is since we're making RuntimeClass as default choice for the users, existing users have to update their pod spec or workload pod template spec to ensure that their workloads are not broken during upgrades. 
This can be mitigated by 
- Having appropriate warnings in the releases before we beta and GA will mitigate the problem.
- Having an example out-of-tree admission webhook which can detect such workloads and update them.(Having an in-tree admission plugin which mutates the pod spec to have appropriate Windows specific RuntimeClass may be too hard to do generically and not transparent to the user)

## Design Details

We can piggyback on the existing RuntimeClass admission controller to query for the RuntimeClass and see if it has
`kubernetes.io/os: windows` in it's scheduling field. 

> TODO: We need to get an ack from sig-auth to use it.

Reviewer 5: Adding them as reviewers and asking for an lgtm should suffice instead of the TODO

Author: I realized that runtimeclasses is owned by sig-node, not sig-auth. So, I need to add them. Will update shortly.

> If we cannot piggyback on the RuntimeClass controller we need to build our own validating admission controller for Windows specific pods.

Reviewer 5: You would still need a way to identify Windows Pods that is not controlled by a regular user as you have called out in the alternatives section.

Author: The validating admission controller does identification + validation because for validation we need identification of Windows pods :)

Reviewer 5: My comment was to further support using `RuntimeClasses` to authoritatively identify Windows Pods.

### Test Plan

Whatever path we choose, unit and e2e tests are a hard requirement to ensure smooth
transition to make RuntimeClasses default choice for the users.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled
- Users getting a warning to switch to RuntimeClasses in beta release

#### Alpha -> Beta Graduation
- Gather feedback from end users
- Tests are in Testgrid and linked in the KEP
- Deny admission to pods that have only `kubernetes.io/os: windows` in nodeSelector
  but don't have an associated RuntimeClass with `kubernetes.io/os: windows` in
  scheduling field.

#### Alpha -> Beta Graduation
- 2 examples of end users using this field

### Upgrade / Downgrade Strategy

- Upgrades:
  When upgrading from a release without this feature, to a release with `RuntimeClassesForWindows` enabled, we will honor both nodeSelector and RuntimeClasses

>   but a warning will be sent to the user that `honoring explicit nodeSelector without Runtimeclasses will be stopped from next release` if the pod has just nodeSelector set. This ensures users have enough time to move without breaking their workloads.

Reviewer 5: Please call out how this warning will be sent. Event etc.

Author: It'll be sent as a message to the user before [persisting](https://github.com/kubernetes/kubernetes/blob/c2882f314218af489267b7163b2ca09c79ffb00a/pkg/registry/apps/daemonset/strategy.go#L170)

- Downgrades:
  When downgrading from a release with this feature to a release without `RuntimeClassesForWindows`, the existing behavior will continue where both nodeSelector and RuntimeClasses are honoured without a warning message.

### Version Skew Strategy

> If the feature gate is enabled, only admission controller(either RuntimeClass or Windows specific admission controller) should be impacted. This feature may have some kubelet implications as the code to strip security constraints based on OS can be removed.

Reviewer 5: This feature will have kubelet implications

Reviewer 5: Should we call out `api-server` here also as per @marosset's [comment](https://github.com/kubernetes/enhancements/pull/2803/commits/3a90d782341954e121941eb747f3918217806c5b#r669944486)?

Reviewer: This enhancement was discussed at sig-node today.
One suggestion make by @derekwaynecarr was it might make more sense to strip the security constraints for that don't apply to the intended OS at admission time rather than have the kubelet remove the fields. That way we wouldn't be persisting unnecessary data and also could help backwards compatibility.

Reviewer 4: i am strongly in favor of making it clear that a pod is authoritatively a windows or linux pod.
i am not sure i like a strong correlation to runtime class as the mechanism for end-users.
i still prefer the idea of a dedicated field on the pod, and regret we did not do what oci spec did by having a dedicated linux/windows section in the pod spec.  it makes it less cumbersome just looking at the pod resource what type of pod this pod is intending to produce.  i also would like to couple kubelet feature support specific to windows with a dedicated field for things that are not supported on windows (ephemeral containers, emptyDir, etc.) that could be supported potentially in the future.  right now, we just admit these pods but fail at runtime is my understanding.

can we add a specific section discussing the negative of an explicit field in the pod spec?

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: RuntimeClassesForWindows

Reviewer 5: I would rather we go with a different name. `RuntimeClassesForWindows` seems to imply that this gate is needed to use `RuntimeClasses` for Windows which is not the case.

Reviewer: Agree that `RuntimeClassesForWindows` can be misleading.
Also, should we try and make this enhancement more os-agnostic but commit to supporting windows checks? 
If so maybe `ConsiderOSDuringAdmission` or something more generic might make more sense?

Author: I don't have any issues with whatever name you folks come up with  :)

Reviewer 5: `OSIdentificationDuringPodAdmission`?

  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

### Scalability

###### Will enabling / using this feature result in any new API calls?

###### Will enabling / using this feature result in introducing new API types?

###### Will enabling / using this feature result in any new calls to the cloud provider?

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

###### What are other known failure modes?

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

## Drawbacks

## Alternatives

> Have a OS specific field in the pod spec which can be either `windows` or `linux` etc. The main advantage of this approach is this can be non-breaking to the existing users, however we already have a way to distinguish pods using the scheduling field in RuntimeClasses. 

Reviewer 5: I would view this as redundant as `RuntimeClasses` provide a similar identification.

Following are the reasons to not choose this approach:
- Having two different ways to do the same thing may be confusing from user experience standpoint
- With proper phasing from using just nodeSelector to RuntimeClasses, we can ensure that the users are not broken. We will also provide an example documentation of this needs to be

> fixed for existing workload and an example admission webhook which can properly mutate the pod spec if a Windows specific RuntimeClass already exists.

Reviewer 5: I would also add that `RuntimeClasses` are needed if we want to run multiple Windows OS variants as worker nodes.

Author: That is questionable. You may still go with combination of nodeSelector and tolerations and os fields without using runtimeclasses

Reviewer 5: Sure, but `RuntimeClasses` seems more elegant. But I agree there are alternatives so I will retract my comment.

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The main goal of this enhancement is to authoritatively identify Windows
pods during the API Server admission time. 

## Motivation

Identifying Windows pods during the API Server admission time is crucial to apply
appropriate security constraints to the pod. Without proper identification, some admission
plugins may apply unnecessary security constraints to the Windows pods or in the
worst case, don't apply security constraints at all.

### Goals

Admission plugins and/or Admission Webhooks identifying if the incoming pod
is Windows or not.

### Non-Goals

- Validating if the Windows pods have appropriate security constraints.
- Interaction with container runtimes.

## Proposal

As mentioned [earlier](#Motivation) identifying the Windows pods during the scheduling phase can be done
in the following ways:

> - Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 

Reviewer 2: I don't think tolerations can be used for this... they expand the pool of possible nodes, but don't restrict the pool to only nodes with particular taints

Reviewer 6: as far as i understand that's why a combination of both node-selector and tolerations is needed.

  Windows node specific labels.

> - Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec

Reviewer 2: which mechanism is this proposing be treated as authoritative? It would be helpful to describe the exact steps/algorithm a pod admission plugin should perform on the data in the incoming pod to determine the target OS

The problem with using nodeSelector and tolerations is that any unprivileged user/entity can apply the 
nodeSelector and/or tolerations in the pod spec making it unsecure whereas RuntimeClasses are [recommended](https://kubernetes.io/docs/concepts/containers/runtime-class/#setup) to be
created by cluster administrator making them authoritative enough to be used.

> In order to make `kubernetes.io/os` authoritative enough, kubelet should also reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value not matching `runtime.GOOS`. We can use this check as `kubernetes.io/os` value is supposed to be filled from `runtime.GOOS`. Having this check in kubelet helps in the following conditions:

Reviewer 2: typo: `kubernetes.io/os`

- If scheduler has been by passed by setting the nodeName directly on the pod spec.
- A malicious user or a user unknowingly with RBAC to just bind pod to node, can associate pod to node there by bypassing the scheduler.

### User Stories (Optional)

#### Story 1

As a Kubernetes cluster administrator, I want appropriate security contexts to be 
applied to my Windows Pods along with Linux pods

#### Story 2

As a Kubernetes cluster administrator, I want to use my own admission webhook for
Windows pods.

### Notes/Constraints/Caveats (Optional)

While the `scheduling` field in the RuntimeClass is created to handle scheduling constraints, we have a reserved label
[`kubernetes.io/os`](https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetes-io-os) to identify
Windows nodes from their Linux counterparts. If the RuntimClass's scheduling field has a nodeSelector  `kubernetes.io/os: 'windows'`,
we can authoritatively say that the pod is indeed a Windows one during the api-server admission time. Querying the RuntimeClass
during admission time may be expensive operation as the admission plugin is not stateless anymore. 

### Risks and Mitigations

> Existing users may be impacted as we're enforcing to use RuntimeClasses instead of plain nodeSelector and tolerations.

Reviewer 2: breaking existing successful Pod API users in this way doesn't match my expectations for GA features

Reviewer 2: specifically, this proposed change does not seem necessary or backwards compatible:

> Deny admission to pods that have only `kubernetes.io/os: windows` in nodeSelector but don't have an associated RuntimeClass with `kubernetes.io/os: windows` in scheduling field.

why is it necessary to reject such a pod? won't it get scheduled and run correctly?

Author: One scenario, that I noticed frequently is users specify a wrong nodeSelector, for example provide windows nodeSelector on linux host and vice-versa. If it is created via a workload controller which creates n replicas, since kubelet would allow such pods to be run the container runtime is going to throw the error and chances of node going unready are high. I just wanted to make sure that we're failing early.

Having said that, if the kubelet can rely on the nodeSelector of the pod and compare it with runtime.GOOS to admit/deny the pod, I think we are in a better shape than we were earlier and I believe it is ok to let users specify the nodeSelector and be less prescriptive.

Reviewer 4: we would still have that pod in a constant bind->admit->reject loop, right?

similar to @liggitt though, i do not know how we can require usage of a runtime class.  for example, if a cluster had windows only nodes, control plane externalized, wouldn't this break backward compatibility for that cluster user?  who is responsible for creation of the runtime class?

> The downside of this approach is since we're making RuntimeClass as default choice for the users, existing users have to update their pod spec or workload pod template spec to ensure that their workloads are not broken during upgrades. 

Reviewer 4: will we document a prescriptive RuntimeClass?  given RuntimeClass usage varies in the wild, this seems to make all windows workload pod examples hard to follow as they are less portable which is why I keep thinking having a dedicated field on the pod spec itself more portable.

This can be mitigated by 
- Having appropriate warnings in the releases before we beta and GA will mitigate the problem.
- Having an example out-of-tree admission webhook which can detect such workloads and update them.(Having an in-tree admission plugin which mutates the pod spec to have appropriate Windows specific RuntimeClass may be too hard to do generically and not transparent to the user)

## Design Details

> We can piggyback on the existing RuntimeClass admission controller to query for the RuntimeClass and see if it has

Reviewer 2: it's unclear what this is proposing the existing RuntimeClass admission plugin do... forbid pods which set a `kubernetes.io/os: windows` spec.nodeSelector and don't set a runtimeClassName?

Author: I can update this section to make it clear.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The main goal of this enhancement is to authoritatively identify the Operating System
on which a pod would like to run.

## Motivation

Identifying the OS of the pods during the API Server admission time is crucial to apply
appropriate security constraints to the pod. Without proper identification, some admission

> plugins may apply unnecessary security constraints to the pods or in the

Reviewer: ```suggestion
plugins may apply unnecessary security constraints to the pods (which may prevent them from running properly) or in the
```

worst case, don't apply security constraints at all.

### Goals

Admission plugins and/or Admission Webhooks identifying the Operating System on which
pod intends to run.

### Non-Goals

> - Validating if the pods have appropriate security constraints.

Reviewer 7: I think the kubelet's behavior of stripping fields should be modified to take into account the new OS field if it is present. If it is not present, then the previous kubelet behavior is appropriate.

The validation performed in the kube-apiserver also needs to be adjusted at the same time we introduce the field.  I suggest enumerating every field to avoid missing one and perhaps writing a fuzzing test of some kind to catch new additions.

> - Validating if the pods have appropriate security constraints.

Author: Fair enough. I was intending to add the API validation anyways. I was thinking Pod API validation and kubelet stripping should happen in the same release and I was pushing back on that.

I misunderstood Jordan's intention here. If he is saying the samething that you're mentioning.

> - Validating if the pods have appropriate security constraints.

Reviewer 2: > The main goal in this release would be to identify the Windows Pods and let the PodSecurityAdmission plugin look at the OS field and ignore such pods in the first release is what I am thinking if doesn't violate existing behaviour.

Any API validation changes that are based on the new OS field should be present in the first release that includes the OS field.

- Interaction with container runtimes.

## Proposal

> Identifying the Windows pods during the scheduling phase can be done

Reviewer 7: This is "pods targeting windows nodes", correct?  I want to be sure the distinction between the OS of the host and OS of the pod remains clear to me while reading this KEP.

in the following ways:
- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
  Windows node specific labels.
- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec

> The runtimeclass is a higher level abstraction which gets translated again to nodeSelectors+tolerations. While the 

Reviewer 7: Does the scheduler honor runtimeclasses today?

Reviewer: Runtime classes usually container nodeSelectors and any nodeSelectors on a runtimeClass do get honored by the scheduler today.

nodeSelector with reserved OS label is good enough, it can be viewed as an side-effect of scheduling constraints 

> and need not necessarily express the user intention fully. So, we propose to add a new field to the pod spec called 

Reviewer 7: I think this field was primarily the result of the distinction between
1. pod targets a windows node
2. pod OS is [linux,windows]

Since a pod targeting a windows node could still be a linux pod, we needed a way to express scheduling distinct from the pod OS.  The pod OS informs what combination of fields can be specified and what values need to be enforced during pod security admission.

os to identify the OS on which pod intends to run.

```go
type PodSpec struct {
  // If specified, indicates the type of OS on which pod intends
  // to run.  This field is alpha-level and is only honored by servers that 
  // enable the IdentifyOS feature. This is used to help identifying the
  // OS authoritatively during the API server admission.
  // +optional
  OS *OS
}

// OS has information on the type of OS. 
// We're making this a struct for possible future expansion.

> type OS struct {

Reviewer 7: There was a fair amount of discussion related to minor versions of windows as specific targets.  As an API consumer, I'd really like to assume all my windows pods have the same fields valid, but I don't actually know.  @liggitt?

Reviewer: I'm OK with treating all versions of Windows as the same for admission/validation purposes.
Almost all functionality that is only supported on some versions of Windows is in networking (like IPv4/IPv6 dualstack support) and these behaviors are usually controlled by kube-proxy flags or Service specs not pod specs.
@jsturtevant @jayunit100 @claudiubelu do any of you have concerns here?

Author: I am resolving the comment here. @deads2k if your comment or question is unanswered please let me know.

  // Name of the OS. The values supported are available at:

>   // https://github.com/opencontainers/runtime-spec/blob/master/config.md#platform-specific-configuration

Reviewer 7: We support all of these?  solaris?

When the API is created, I think a strict enumeration is required.

Reviewer 2: current node platforms:
```
# The node platforms we build for
readonly KUBE_SUPPORTED_NODE_PLATFORMS=(
  linux/amd64
  linux/arm
  linux/arm64
  linux/s390x
  linux/ppc64le
  windows/amd64
)
```

Reviewer 2: that said, I'm not necessarily opposed to accepting more OS values, I just think that means more work to define API validation and PodSecurity treatment of each enumerated value

when we add validation for this, we should probably bake in toleration of an existing unknown value on an update, and explicitly note that future values could be added in the future, and how a client should handle an unknown enum value (xref https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field)

Author: Ahh I see, I was doing something like this in the validation stage - https://github.com/kubernetes/kubernetes/pull/104693/commits/852f44253137ddfba8209eee1385827f9657f835#diff-c713e8919642d873fdf48fe8fb6d43e5cb2f53fd601066ff53580ea655948f0dR6158. I believe I need to add some toleration for the unknown updates.

  Name string
}

```

> With the above change, all the admission plugins which validate or mutate the pod can 

Reviewer 7: I think the API validators like ValidatePod and ValidatePodUpdate can become more strict as well.  Validation for a linux pod could require that windows fields be empty.  Validation for a windows pod could require that linux fields be empty.

have a validating admission plugin for Windows pods as Linux and Windows host capabilities are
different and not all Kubernetes features are supported on Windows host.

### User Stories (Optional)

#### Story 1

As a Kubernetes cluster administrator, I want appropriate security contexts to be 
applied to my Windows Pods along with Linux pods

#### Story 2

As a Kubernetes cluster administrator, I want to use my own admission webhook for pods
based on the OS it intends to run on.

### Notes/Constraints/Caveats (Optional)

Since the OS field is optional in the pod spec, we can expect the users to omit this field
when submitting pod to API server. In this scenario, we expect admission plugins to treat the pod
the way it is being treated now.

### Risks and Mitigations

The primary risk is a bug in the implementation of the admission plugins that validate or mutate
based on the OS field in the pod spec. The best mitigation for that scenario is unit testing 
when this featuregate is enabled and disabled.

## Design Details

### Implications to Kubelet

Apart from the above API change, we intend to make the following changes to Kubelet:

> - Kubelet should reject admitting pods if the pod.Spec.OS.Name doesn't match the node label `kubernetes.io/os`

Reviewer 7: When/if linux containers on windows are supported, this will need to be updated, correct?  I think it's still ok to add this since old kubelets won't be able to correctly manage such pods, but I want to call out our future debt.

Perhaps it is better phrased as, "kubelets should reject admitting pod if the kubelet cannot honor the pod.spec.os.name. For instance, if the os.name does not match the host os".

Reviewer: Agree, matching the `kubernetes.io/os` node label is an implementation detail. For the KEP we can use the phasing mentioned above.

- Kubelet should reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value 
not matching `runtime.GOOS`. We can use this check as `kubernetes.io/os` value is supposed to be filled from `runtime.GOOS`. 

Having the above checks in kubelet helps in the following conditions:
- If scheduler has been by passed by setting the nodeName directly on the pod spec.
- A malicious user or a user unknowingly with RBAC to just bind pod to node, can associate pod to node there by bypassing the scheduler.

As of now, Kubelet doesn't reconcile the value of `kubernetes.io/os`. So, it'd be nice to have the reconciliation as part of this change.

### Implications to Scheduler

We let the users to explicitly specify nodeSelectors+tolerations or runtimeclasses to express their intention

> to run an particular OS. However, in future, once the OS struct expands, we can see if we can leverage those fields to

Reviewer 7: I think such a statement of intent requires signoff from sig-scheduling or deferral from this KEP.

Author: Fair enough. I am tagging scheduling folks. 

cc @kubernetes/sig-scheduling-misc @Huang-Wei @ahg-g @damemi

Reviewer 8: if we just soften the statement, can we push forward *without* sig-scheduling signoff :), after all i dont think were actually planning on modifying the scheduler here are we?

Reviewer 9: @ravisantoshgudimetla LGTM on scheduling side.

express scheduling constraints. During the alpha, we assume there are no scheduling implications.

### Test Plan

- Unit tests covering API server defaulting to various fields within pod with and without this feature
- Unit tests covering admission plugins which validate/mutate pod spec based on this feature
- Unit and E2E tests for Kubelet changes.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag
- Basic units and e2e tests completed and enabled

#### Alpha -> Beta Graduation
- Expand the e2e tests with more scenarios
- Gather feedback from end users
- Tests are in Testgrid and linked in the KEP

#### Alpha -> Beta Graduation
- 2 examples of end users using this field

### Upgrade / Downgrade Strategy

- Upgrades:
  When upgrading from a release without this feature, to a release with `IdentifyOS` feature enabled, there is no change to existing behavior unless user specifies
  OS field in the pod spec.
- Downgrades:
  When downgrading from a release with this feature to a release without `IdentifyOS`, the current behavior will continue.

> ### Version Skew Strategy

Reviewer 7: I'd like to see a discussion of the how an old kubelet interprets new pods.

Author: Addressed it below

If the feature gate is enabled there are some kubelet implications as the code to strip security constraints based on OS can be removed and we need to add
admission/denying in the kubelet logic which was mentioned above.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: IdentifyOS

Reviewer 7: I prefer IdentifyPodOS

Author: I don't have a strong preference. If that is the name everyone is fine with, I can use that name.

  - Components depending on the feature gate:
    - kubelet
    - kube-apiserver

###### Does enabling the feature change any default behavior?
No

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes. Using the featuregate is the only way to enable/disable this feature

Reviewer 7: if pod validation changes to allow or require some fields to be empty, the update validation will need to need to be adjusted to accommodate missing fields or pods will become un-updateable.  I'd to explicitly call out the need for unit tests and manual tests ensuring this compatibility.

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> The main goal of this enhancement is to authoritatively identify the Operating System

Reviewer 8: I feel like "would like to run" is imprecise, b/c the goal is to talk about `enforcement`, right?
```suggestion
The primary goal of this enhancement is to define the OperatingSystem scheduling requirements for a Pod in a manner which is common in the Kubelet as well as the KubernetesControlPlane.
```

on which a pod would like to run.

## Motivation

Identifying the OS of the pods during the API Server admission time is crucial to apply

> appropriate security constraints to the pod. Without proper identification, some admission

Reviewer 8: minor clarification to the `proper identification` clause... 
```suggestion
appropriate security constraints to the pod. In the absence of having information about a target OS for a pod, some admission
```

plugins may apply unnecessary security constraints to the pods(which may prevent them from running properly) or in the
worst case, don't apply security constraints at all.

### Goals

> Admission plugins and/or Admission Webhooks identifying the Operating System on which

Reviewer 8: I think we mean to standardize this webhook into Kubernetes/core itself, right?
```suggestion
To add a standard admission plugins and/or an admission webhooks to Kubernetes which can identify the Operating System on which a pod should run on
```

Author: Nope, the admission plugins can be outside of core Kube too.

pod intends to run.

### Non-Goals

- Interaction with container runtimes.

> - Interaction with scheduler. (This may change in future)

Reviewer 8: ```suggestion
- Interaction with scheduler (this may change in future).
```

## Proposal

Identifying the pods targeting Windows nodes during the scheduling phase can be done
in the following ways:
- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
  Windows node specific labels.
- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec

The runtimeclass is a higher level abstraction which gets translated again to nodeSelectors+tolerations. While the 
nodeSelector with reserved OS label is good enough, it has following shortcomings:

- It can be viewed as an side-effect of scheduling constraints 
and need not necessarily express the user intention fully.
- The pod may target Windows node but pod OS may be linux or windows
as Linux Containers on Windows(LCOW) can be supported in future.

So, we propose to add a new field to the pod spec called 

> os to identify the OS on which pod intends to run.

Reviewer 2: maybe it's just me, but it's not clear if "the OS on which pod intends to run" refers to the node OS or the container OS. "the OS of the containers specified in the pod" would be clearer.

Author: But it is not at the container field level in the pod right? It'd just be at the pod level?

```go
type PodSpec struct {
  // If specified, indicates the type of OS on which pod intends

>   // to run.  This field is alpha-level and is only honored by servers that 

Author: The intention here is that the feature flag is at the api-server level.

  // enable the IdentifyOS feature. This is used to help identifying the
  // OS authoritatively during the API server admission.

>   // +optional

Reviewer: Should we add a note that there will be no default value here?

Author: Done

plugin is the only admission plugin can leverage this field on the pod spec. In addition, API validators like ValidatePod and ValidaPodUpdate should be modified. In future, we can
have a validating admission plugin for Windows pods as Linux and Windows host capabilities are
different and not all Kubernetes features are supported on Windows host.

### User Stories (Optional)

#### Story 1

As a Kubernetes cluster administrator, I want appropriate security contexts to be 
applied to my Windows Pods along with Linux pods

#### Story 2

As a Kubernetes cluster administrator, I want to use my own admission webhook for pods
based on the OS it intends to run on.

### Notes/Constraints/Caveats (Optional)

Since the OS field is optional in the pod spec, we can expect the users to omit this field
when submitting pod to API server. In this scenario, we expect admission plugins to treat the pod
the way it is being treated now.

### Risks and Mitigations

The primary risk is a bug in the implementation of the admission plugins that validate or mutate
based on the OS field in the pod spec. The best mitigation for that scenario is unit testing 
when this featuregate is enabled and disabled.

> 

Reviewer 8: ```suggestion
Additionally, there may be some end-user confusion on the functional consequences of setting the new OS field, given that it is optional.
```

## Design Details

> ### Implications to Kubelet

Reviewer 8: since we're adding kubelet logic, we should just say so :)
```suggestion
### Changes to Kubelet
```

Apart from the above API change, we intend to make the following changes to Kubelet:
- Kubelet should reject admitting pod if the kubelet cannot honor the pod.Spec.OS.Name. For instance, if the OS.Name does not match the host os.
- Kubelet should reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value 
not matching `runtime.GOOS`. We can use this check as `kubernetes.io/os` value is supposed to be filled from `runtime.GOOS`. 

Having the above checks in kubelet helps in the following conditions:
- If scheduler has been by passed by setting the nodeName directly on the pod spec.
- A malicious user or a user unknowingly with RBAC to just bind pod to node, can associate pod to node there by bypassing the scheduler.

As of now, Kubelet doesn't reconcile the value of `kubernetes.io/os`. So, it'd be nice to have the reconciliation as part of this change.

> ### Implications to Scheduler

Reviewer 8: since were not forcing an immediate change ...
```suggestion
### Potential future changes to Scheduler
```

> We let the users to explicitly specify nodeSelectors+tolerations or runtimeclasses to express their intention

Reviewer 9: ```suggestion
We let the users to explicitly specify nodeSelectors/nodeAffinities+tolerations or runtimeclasses to express their intention
```

to run an particular OS. However, in future, once the OS struct expands, we can see if we can leverage those fields to
express scheduling constraints. During the alpha, we assume there are no scheduling implications.

### Test Plan

- Unit tests covering API server defaulting to various fields within pod with and without this feature
- Unit tests covering admission plugins which validate/mutate pod spec based on this feature
- Unit and E2E tests for Kubelet changes.

> 

Reviewer 8: I guess we can clean up the windows tests once this feature goes in?
```suggestion
- Updates to the `sig-windows` tagged tests to utilize to direct windows scheduling for all pods .
```

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag
- Basic units and e2e tests completed and enabled

#### Alpha -> Beta Graduation
- Expand the e2e tests with more scenarios
- Gather feedback from end users
- Tests are in Testgrid and linked in the KEP

> #### Alpha -> Beta Graduation

Reviewer 9: Beta -> GA?

> - 2 examples of end users using this field

Reviewer 8: ```suggestion
- end to end tests confirmed as passing, with the new label, on old as well as new APIServer/Kubelet versions (i.e. proof of backwards compatibilty) 
```

Author: What label are you referring to here?

Reviewer 8: s/label/thing that we use to identify the host OS/g :)

### Upgrade / Downgrade Strategy

- Upgrades:
  When upgrading from a release without this feature, to a release with `IdentifyOS` feature enabled, there is no change to existing behavior unless user specifies
  OS field in the pod spec.
- Downgrades:
  When downgrading from a release with this feature to a release without `IdentifyOS`, the current behavior will continue.
### Version Skew Strategy

If the feature gate is enabled there are some kubelet implications as the code to strip security constraints based on OS can be removed and we need to add
admission/denying in the kubelet logic which was mentioned above. Older Kubelets without this change will continue stripping the unnecessary fields in the pod spec which is the current behavior.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: IdentifyOS
  - Components depending on the feature gate:
    - kubelet
    - kube-apiserver

###### Does enabling the feature change any default behavior?

> No

Reviewer 8: ```suggestion
A Kubelet with a misscheduled pod (i.e. trying to run a windows pod on a linux node) will fail *before* trying to run a container, as opposed to after.
```

Reviewer 8: (@ravisantoshgudimetla keep me honest here in case im missing something)

Reviewer 7: > (@ravisantoshgudimetla keep me honest here in case im missing something)

That's how I understand the kubelet change 

> Kubelet should reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value 
not matching `runtime.GOOS`

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

###### What happens if we reenable the feature if it was previously rolled back?

The admission plugins/Kubelet can act based on the OS field in pod spec if set by the end-user

###### Are there any tests for feature enablement/disablement?
Yes, unit and e2es tests for feature enabled, disabled

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?
It shouldn't impact already running workloads. This is an opt-in feature since users need to explicitly set the OS parameter in the Pod spec i.e .spec.os field. 
If the feature is disabled the field is preserved if it was already set in the presisted pod object, otherwise it is silently dropped.

> ###### What specific metrics should inform a rollback?

Reviewer 7: seems like pod creation and pod update rejections are worth watching.

Author: Agreed and added a metric which we can use to track the pod status however I couldn't find rejections from the admission controller rejection metric unless I misunderstood you. 

I added a metric which can benefit us - https://github.com/kubernetes/enhancements/pull/2803/files#diff-0bd86b6b4704c4e10a453c7f15e74e31aacb53891087cf069ae11b8bf5293a30R320

Author: @deads2k suggested that I can use 400 series errors to capture the rejection at the admission plugin. So, we won't need a new metric. I mentioned the metric we can use to track the 400 series errors [here](https://github.com/kubernetes/enhancements/pull/2803/files#diff-0bd86b6b4704c4e10a453c7f15e74e31aacb53891087cf069ae11b8bf5293a30R316)

N/A

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
Not yet tested.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
None. This feature will be additive

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

By looking at the pod's spec. Some fields like Security Constraint Contexts may not
be applied to the pod spec if the proper OS has been set.

###### How can someone using this feature know that it is working for their instance?

Windows Pods and Linux Pods with proper OS field set in the pod spec would cause pods
to run properly on desired OS

- [x] Events
  - Event Reason: Kubelet will send pod admitted/denied event.
- [ ] API .status
  - Condition name: 
  - Other field: 
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

> All the pods which have OS field set in the pod spec should have OS specific constraints/defaults applied.

Reviewer 8: hm, is this a SLO? i dont know if we have any SLOs maybe just delete it?

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

> - [ ] Metrics

Reviewer 7: The most likely failure mode is rejection either by the kube-apiserver on pod create/update or by the kubelet during kubelet admission.  I suspect both of those have metrics associated them.  I think you should find those metrics and list them.

  - Metric name:
  - [Optional] Aggregation method:
  - Components exposing the metric:
- [x] Other (treat as last resort)
  - Details: Linux Specific SCCs not available on Windows pod specs and vice-versa

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

> None

Reviewer 7: kube-apiserver and kubelet, right?

It's important to list these because a kubelet is often mismatched to a kube-apiserver and so the change won't fully function until kubelets are updated.

### Scalability

###### Will enabling / using this feature result in any new API calls?
No

###### Will enabling / using this feature result in introducing new API types?
Yes

###### Will enabling / using this feature result in any new calls to the cloud provider?
No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?
It increases the size of Pod object since a new string field is added.

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
No

### Troubleshooting

> ###### How does this feature react if the API server and/or etcd is unavailable?

Reviewer 8: i guess we need to fill these out?...

Reviewer 7: > i guess we need to fill these next several q's out?...

he deleted the template comments, but technically only for GA I think.

Author: That's correct. We can fill these questions when we GA this feature.

pod intends to run.

### Non-Goals

- Interaction with container runtimes.
- Interaction with scheduler (this may change in future).

## Proposal

Identifying the pods targeting Windows nodes during the scheduling phase can be done
in the following ways:
- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
  Windows node specific labels.
- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec

The runtimeclass is a higher level abstraction which gets translated again to nodeSelectors+tolerations. While the 
nodeSelector with reserved OS label is good enough, it has following shortcomings:

- It can be viewed as an side-effect of scheduling constraints 
and need not necessarily express the user intention fully.
- The pod may target Windows node but pod OS may be linux or windows

> as Linux Containers on Windows(LCOW) can be supported in future.

Reviewer 2: These sections should be moved to an "alternatives considered" section to make the proposal section succinct.

Author: I've moved those lines to `Alternatives considered` section.

> 

Reviewer 2: the actual proposal starts here

have a validating admission plugin for Windows pods as Linux and Windows host capabilities are
different and not all Kubernetes features are supported on Windows host.

### User Stories (Optional)

#### Story 1

As a Kubernetes cluster administrator, I want appropriate security contexts to be 
applied to my Windows Pods along with Linux pods

#### Story 2

As a Kubernetes cluster administrator, I want to use my own admission webhook for pods
based on the OS it intends to run on.

### Notes/Constraints/Caveats (Optional)

Since the OS field is optional in the pod spec, we can expect the users to omit this field
when submitting pod to API server. In this scenario, we expect admission plugins to treat the pod
the way it is being treated now.

### Risks and Mitigations

The primary risk is a bug in the implementation of the admission plugins that validate or mutate
based on the OS field in the pod spec. The best mitigation for that scenario is unit testing 
when this featuregate is enabled and disabled.
Additionally, there may be some end-user confusion on the functional consequences of setting the new OS field, given that it is optional.

> ## Design Details

Reviewer 2: need sections detailing changes to kube-apiserver and pod security standards (I'd suggest the list from https://github.com/kubernetes/enhancements/pull/2803#discussion_r695196825)

For each accepted OS value:

* Pod spec API validation will be adjusted to ensure values are not set for OS-specific fields that are irrelevant to the Pod's OS.
* Unit tests will be added to ensure new fields in the pod spec are classified as OS-specific or not (and which OSes they are allowed for)

* Pod Security Standards will be reviewed and updated to indicate which Pod OSes they apply to
* The `restricted` Pod Security Standard will be reviewed to see if there are OS-specific requirements that should be added
* The PodSecurity admission implementation will be updated to skip checks which do not apply to the Pod's OS.

Author: Included the changes from comments in separate sections. I also added some text around e2e [below](https://github.com/kubernetes/enhancements/pull/2803/files#diff-0bd86b6b4704c4e10a453c7f15e74e31aacb53891087cf069ae11b8bf5293a30R154)

Reviewer: +1 for e2e tests demonstrating  functionality on windows nodes

### Changes to Kubelet

Apart from the above API change, we intend to make the following changes to Kubelet:
- Kubelet should reject admitting pod if the kubelet cannot honor the pod.Spec.OS.Name. For instance, if the OS.Name does not match the host os.
- Kubelet should reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value 
not matching `runtime.GOOS`. We can use this check as `kubernetes.io/os` value is supposed to be filled from `runtime.GOOS`. 

Having the above checks in kubelet helps in the following conditions:
- If scheduler has been bypassed by setting the nodeName directly on the pod spec.
- A malicious user or a user unknowingly with RBAC to just bind pod to node, can associate pod to node there by bypassing the scheduler.

> As of now, Kubelet doesn't reconcile the value of `kubernetes.io/os`. So, it'd be nice to have the reconciliation as part of this change.

Reviewer 2: this seems like a fine change to make, but I don't think it's related to this KEP any more (definitively identifying the OS of a pod)

Reviewer 4: I agree with @liggitt that its good for kubelet to reconcile the label, but it is distinct.

The kubelet changes described in this section lgtm.

Author: Removed it altogether
